pymc3/backends/arviz.py

```diff
@@ -0,0 +1,606 @@
+"""PyMC3-ArviZ conversion code."""
+import logging
+import warnings
+
+from typing import (  # pylint: disable=unused-import
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    Iterable,
+    List,
+    Mapping,
+    Optional,
+    Tuple,
+    Union,
+)
+
+import numpy as np
+import xarray as xr
+
+from aesara.gof.graph import ancestors
+from aesara.tensor.var import TensorVariable
+from arviz import InferenceData, concat, rcParams
+from arviz.data.base import CoordSpec, DimSpec, dict_to_dataset, requires
+
+import pymc3
+
+from pymc3.model import modelcontext
+from pymc3.sampling import _DefaultTrace
+from pymc3.util import get_default_varnames
+
+if TYPE_CHECKING:
+    from typing import Set  # pylint: disable=ungrouped-imports
+
+    from pymc3.backends.base import MultiTrace  # pylint: disable=invalid-name
+    from pymc3.model import Model
+
+___all__ = [""]
+
+_log = logging.getLogger("pymc3")
+
+# random variable object ...
+Var = Any  # pylint: disable=invalid-name
+
+
+class InferenceDataConverter:  # pylint: disable=too-many-instance-attributes
+    """Encapsulate InferenceData specific logic."""
+
+    model = None  # type: Optional[Model]
+    nchains = None  # type: int
+    ndraws = None  # type: int
+    posterior_predictive = None  # Type: Optional[Mapping[str, np.ndarray]]
+    predictions = None  # Type: Optional[Mapping[str, np.ndarray]]
+    prior = None  # Type: Optional[Mapping[str, np.ndarray]]
+
+    def __init__(
+        self,
+        *,
+        trace=None,
+        prior=None,
+        posterior_predictive=None,
+        log_likelihood=True,
+        predictions=None,
+        coords: Optional[CoordSpec] = None,
+        dims: Optional[DimSpec] = None,
+        model=None,
+        save_warmup: Optional[bool] = None,
+        density_dist_obs: bool = True,
+        index_origin: Optional[int] = None,
+    ):
+
+        self.save_warmup = rcParams["data.save_warmup"] if save_warmup is None else save_warmup
+        self.trace = trace
+
+        # this permits us to get the model from command-line argument or from with model:
+        self.model = modelcontext(model)
+
+        self.attrs = None
+        if trace is not None:
+            self.nchains = trace.nchains if hasattr(trace, "nchains") else 1
+            if hasattr(trace.report, "n_draws") and trace.report.n_draws is not None:
```

Comment:
@michaelosthege is it possible for traces in v4 to not have a report?

Reply:
I don't think the `v4` changes should affect that, so the answer is most likely "yes".

Reply:
IIRC this `if` was just for backwards compatibility!
With the code joining the PyMC3 codebase, we can just forget about backwards compatibility entirely!
You can double check with `sampling.py`, but I'm pretty sure traces will always have a report.